index.js

```diff
@@ -873,22 +878,25 @@
   //  Object$prototype$concat :: StrMap a ~> StrMap a -> StrMap a
   function Object$prototype$concat(other) {
     var result = {};
-    for (var k in this) result[k] = this[k];
-    for (k in other) result[k] = other[k];
+    function assign(k) { result[k] = this[k]; }
+    forEachKey(this, assign);
+    forEachKey(other, assign);
     return result;
   }
 
   //  Object$prototype$map :: StrMap a ~> (a -> b) -> StrMap b
   function Object$prototype$map(f) {
     var result = {};
-    for (var k in this) result[k] = f(this[k]);
+    forEachKey(this, function(k) { result[k] = f(this[k]); });
     return result;
   }
 
   //  Object$prototype$ap :: StrMap a ~> StrMap (a -> b) -> StrMap b
   function Object$prototype$ap(other) {
     var result = {};
-    for (var k in this) if (k in other) result[k] = other[k](this[k]);
+    forEachKey(this, function(k) {
+      if (has(k, other)) result[k] = other[k](this[k]);
+    });
```

Comment:
Note that we're replacing `k in other` with `has(k, other)` here. `Z.ap({}, {toString: 42})` behaves correctly as a result.